### PR TITLE
The AlluxioFS implementation for better supporting Hadoop 2.x

### PR DESCRIPTION
**Pull request by pfxuan**
Hadoop 2.x uses AbstractFileSystem as the interface to access other under compatible file systems. This commit uses the delegated Alluxio Hadoop 1.x Filesystem as initial implementation.

### CODE REVIEW

File: docs/en/Running-Hadoop-MapReduce-on-Alluxio.md [hunk 1]
```diff
@@ -28,13 +28,13 @@ the `hdfs-site.xml` file as well.
...
-If you are using a 2.x Hadoop cluster, you should not need the properties above in your
-`core-site.xml` file. However, in some cases you may encounter the error:
+If you are using a Hadoop 2.x cluster, you need to configure `alluxio.hadoop.AlluxioFs` as Hadoop AbstractFileSystem. 
+The properties of Hadoop 1.x is optional. However, in some cases you may encounter the error:
 `java.io.IOException: No FileSystem for scheme: alluxio`. For instance, this may happen when Yarn 
 (as opposed to Hadoop) tries to access Alluxio files. If this error is encountered, add these 
 properties to your `core-site.xml` file, and restart Yarn.
 
-{% include Running-Hadoop-MapReduce-on-Alluxio/config-core-site.md %}
+{% include Running-Hadoop-MapReduce-on-Alluxio/config-core-site-hadoop-2.x.md %}
```
Comment (calvinjia) [inline docs/en/Running-Hadoop-MapReduce-on-Alluxio.md]: Is there a way we can word this so the user doesn't need to encounter an error before setting this up, there's no downside to adding the properties right?

Comment (pfxuan) [inline docs/en/Running-Hadoop-MapReduce-on-Alluxio.md]: Yes. This is a good idea! There is no downside for adding this property. Basically, Hadoop 2.x uses two mixed interfaces together. e.g. in Hadoop dfs command line still uses the old FileSystem interface, and Yarn MRv2 requires AbstractFileSystem interface. So we can just one fixed configuration and share it to Hadoop 1.x and Hadoop 2.x.

**Comment by pfxuan** [comment]
@calvinjia, I modified the document part and made Hadoop 1.x and 2.x share the same configuration. Please help check.

File: core/client-internal/src/main/java/alluxio/hadoop/AlluxioFs.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+ * existing Alluxio FileSystem and Only necessary for use with Hadoop 2.x.
+ * Configuration example in Hadoop core-site.xml file:
+ * <pre>
+ * &lt;property&gt;
+ *    &lt;name>fs.AbstractFileSystem.alluxio.impl&lt;/name&gt;
+ *    &lt;value>alluxio.hadoop.AlluxioFs&lt;/value&gt;
+ * &lt;/property&gt;
+ * </pre>
+ */
+public class AlluxioFs extends DelegateToFileSystem {
```
Comment (calvinjia) [inline core/client-internal/src/main/java/alluxio/hadoop/AlluxioFs.java]: Could you link some documentation for extending `DelegateToFileSystem`? It seems like it is a private and unstable interface.

Comment (pfxuan) [inline core/client-internal/src/main/java/alluxio/hadoop/AlluxioFs.java]: While `DelegateToFileSystem` is labeled as a private/unstable interface, this interface is widely used by many other compatible file systems (such as WebHdfs, SwebHdfs, HarFs, FTPfs, RawLocalFs, Cephfs, GlusterFS, NFS and [so on](https://www.google.com/#q=extends+DelegateToFileSystem+%7B+site:github.com)) with [long history](https://github.com/apache/hadoop/commits/trunk/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/DelegateToFileSystem.java). For the long term solution, we can write another AlluxioFs by extending Hadoop `AbstractFileSystem` directly.

**Comment by pfxuan** [comment]
@calvinjia, the document has been linked. Please help to check.

File: core/client-internal/src/main/java/alluxio/hadoop/AlluxioFs.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+  private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
+
+  /**
+   * This constructor has the signature needed by
+   * {@link org.apache.hadoop.fs.AbstractFileSystem#createFileSystem(java.net.URI, Configuration)}.
+   *
+   * @param uri which must be that of AlluxioFS
+   * @param conf
+   * @throws java.io.IOException
+   * @throws java.net.URISyntaxException
```
Comment (calvinjia) [inline core/client-internal/src/main/java/alluxio/hadoop/AlluxioFs.java]: Could you complete the java doc here? also we generally don't include the package elements and just the class name in the throws annotation.

**Comment by calvinjia** [comment]
@pfxuan Thanks for adding the explanation. One minor comment otherwise looks good.

**Comment by pfxuan** [comment]
@calvinjia, all your comments have been addressed. Please help to check.

**Comment by calvinjia** [comment]
LGTM

**Pull request merged**